inc/cappuccino/UtMap.hpp

```diff
@@ -0,0 +1,475 @@
+    /**
+     * @param uniform_ttl The uniform TTL for key values inserted into the map. 100ms default.
+     */
+    UtMap(std::chrono::milliseconds uniform_ttl = std::chrono::milliseconds{100});
+
+    /**
+     * Inserts or updates the given key value pair using the uniform TTL.  On update will 
+     * reset the TTL.
+     * @param key The key to store the value under.
+     * @param value The value of data to store.
+     * @param allow Allowed methods of insertion | update.  Defaults to allowing
+     *              insertions and updates.
+     * @return True if the operation was successful based on `allow`.
+     */
+    auto Insert(
+        const KeyType& key,
+        ValueType value,
+        Allow allow = Allow::INSERT_OR_UPDATE) -> bool;
+
+    /**
+     * Inserts or updates a range of key value pairs using the default TTL.  
+     * This expects a container that has 2 values in the {KeyType, ValueType} 
+     * ordering.  There is a simple struct UtMap::KeyValue that can be put into 
+     * any iterable container to satisfy this requirement.
+     * @tparam RangeType A container with two items, KeyType, ValueType.
+     * @param key_value_range The elements to insert or update into the map.
+     * @param allow Allowed methods of insertion | update.  Defaults to allowing
+     *              insertions and updates.
+     * @return The number of elements inserted based on `allow`.
+     */
+    template <typename RangeType>
+    auto InsertRange(
+        RangeType&& key_value_range,
+        Allow allow = Allow::INSERT_OR_UPDATE) -> size_t;
+
+    /**
+     * Attempts to delete the given key.
+     * @param key The key to remove from the map.
+     * @return True if the key was deleted, false if the key does not exist.
+     */
+    auto Delete(
+        const KeyType& key) -> bool;
+
+    /**
+     * Attempts to delete all given keys.
+     * @tparam RangeType A container with the set of keys to delete, e.g. vector<k> or set<k>.
+     * @param key_range The keys to delete from the map.
+     * @return The number of items deleted from the map.
+     */
+    template <template <class...> typename RangeType>
+    auto DeleteRange(
+        const RangeType<KeyType>& key_range) -> size_t;
+
+    /**
+     * Attempts to find the given key's value.
+     * @param key The key to lookup its value.
+     * @return An optional with the key's value if it exists, or an empty optional if it does not.
+     */
+    auto Find(const KeyType& key) -> std::optional<ValueType>;
+
+    /**
+     * Attempts to find all the given keys values.
+     * @tparam RangeType A container with the set of keys to lookup, e.g. vector<KeyType>.
+     * @param key_range A container with the set of keys to lookup.
+     * @return All input keys to either a std::nullopt if it doesn't exist, or the value if it does.
+     */
+    template <typename RangeType>
+    auto FindRange(const RangeType& key_range) -> std::vector<std::pair<KeyType, std::optional<ValueType>>>;
```

Comment:
I think the reason this is like this, is because the generic RangeType specialization works as long as it can use a range-based for loop. If we want to return a RangeType we would also have to specialize how it is constructed, which is much less generic than looping for containers.

Reply:
I don't think preston is suggesting to return RangeType, I think he means to template specialize the RangeType template parameter like DeleteRange does to enforce the type it contains as  KeyType.  It needs to be a ... template parameter though to allow for special allocators (something we are not using but theoretically could).